pkg/frontend/querymiddleware/cardinality_test.go

```diff
@@ -0,0 +1,341 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/gogo/protobuf/proto"
+	"github.com/grafana/dskit/cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/user"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/util"
+)
+
+func Test_cardinalityEstimateBucket_GenerateCacheKey_keyFormat(t *testing.T) {
+	requestTime := parseTimeRFC3339(t, "2023-01-09T03:24:12Z")
+	hoursSinceEpoch := util.TimeToMillis(requestTime) / time.Hour.Milliseconds()
+	daysSinceEpoch := hoursSinceEpoch / 24
+
+	type args struct {
+		userID string
+		r      Request
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		{
+			name: "instant query",
+			args: args{
+				userID: "tenant-a",
+				r: &PrometheusInstantQueryRequest{
+					Time:  requestTime.UnixMilli(),
+					Query: "up",
+				},
+			},
+			want: fmt.Sprintf("tenant-a:up:%d", daysSinceEpoch),
+		},
+		{
+			name: "range query",
+			args: args{
+				userID: "tenant-b",
+				r: &PrometheusRangeQueryRequest{
+					Start: requestTime.UnixMilli(),
+					End:   requestTime.Add(2 * time.Hour).UnixMilli(),
+					Query: "up",
+				},
+			},
+			want: fmt.Sprintf("tenant-b:up:%d", daysSinceEpoch),
+		},
+		{
+			name: "range query with large range",
+			args: args{
+				userID: "tenant-b",
+				r: &PrometheusRangeQueryRequest{
+					Start: requestTime.UnixMilli(),
+					// Over 24 hours, should add range part to key
+					End:   requestTime.Add(25 * time.Hour).UnixMilli(),
+					Query: "up",
+				},
+			},
+			want: fmt.Sprintf("tenant-b:up:%d:%d", daysSinceEpoch, 1),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			assert.Equal(t, tt.want, generateCacheKey(tt.args.userID, tt.args.r, 24*time.Hour))
+		})
+	}
+}
+
+func Test_cardinalityEstimation_lookupCardinalityForKey(t *testing.T) {
+	ctx := context.Background()
+	c := cache.NewInstrumentedMockCache()
+
+	presentKey := "tenant-a:up:1234:4321"
+	presentValue := uint64(25)
+	marshaled, err := proto.Marshal(&QueryCardinality{presentValue})
+	require.NoError(t, err)
+	c.Store(ctx, map[string][]byte{cacheHashKey(presentKey): marshaled}, 1*time.Hour)
```

Comment:
I would directly use `storeCardinalityForKey()` so you exercise it too.

Reply:
I chose to store here using the cache  directly because `storeCardinalityForKey` is not available without the struct and this data is part of the fixture for this test. `storeCardinalityForKey` is covered by the test though, do you think it'd be worth adding a dedicated test for that method?

Reply:
Ah right, this is true for this location, will change it. My comment above was meant for the test targeting `Do`, because there the struct isn't available at all.